### PR TITLE
Add graphviz model graphs

### PR DESCRIPTION
**Pull request by ColCarroll**
This follows #1683 and #876 to create model representations for graphs.  I also reran the [multilevel modeling](https://docs.pymc.io/notebooks/multilevel_modeling.html) notebook, updating a bit of the code, and adding the graphs for each of the models.

Installation of graphviz is now supported on conda-forge, and is used in `dask`, so it seems like a reasonable choice.  I've also isolated the code, so it is easy to remove if it causes headaches.

The function takes only one (optional) argument - the model object, which it can graph from the context manager if necessary.  It returns a `graphviz.Digraph`, so if someone wanted to customize the object for publication or any other reason, that is surprisingly easy.  My favorite part is that the graphs display inline in a notebook.

I've included two examples of the graphs from the Radon example, since they look interesting, and show all of the styling I added:

- Rectangles, ovals, shaded ovals for deterministic, RV, and observed variables
- Plate detection (I just group by variable shape, so it can definitely be wrong)
- A subtle rounded corner on the plates for a touch of class

image

image

### CODE REVIEW

File: pymc3/model_graph.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    Requires graphviz, which may be installed most easily with
+        conda install -c conda-forge python-graphviz
+
+    Alternatively, you may install the `graphviz` binaries yourself,
+    and then `pip install graphviz` to get the python bindings.  See
+    http://graphviz.readthedocs.io/en/stable/manual.html
+    for more information.
+    """
+    model = pm.modelcontext(model)
+    return ModelGraph(model).make_graph()
```
Comment (fonnesbeck) [inline pymc3/model_graph.py:174]: Need newline

File: pymc3/model_graph.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        return only the inputs *to the deterministic*.  However, if we pass in the
+        deterministic as a blocker, it will skip those nodes.
+        """
+        deterministics = self.get_deterministics(var)
+        upstream = self._inputs(var, func)
+        parents = self._inputs(var, func, blockers=deterministics)
+        if parents != upstream:
+            det_map = {}
+            for d in deterministics:
+                d_set = set([j for j in inputs([func], blockers=[d])])
```
Comment (fonnesbeck) [inline pymc3/model_graph.py]: Can just use a set comprehension: `{j for j in inputs([func], blockers=[d])}`

**Comment by fonnesbeck** [comment]
This is great! Can we just make this the repr for `Model`?

**Comment by rlouf** [comment]
This looks very cool! Have you considered daft?

**Comment by junpenglao** [comment]
This is awesome! Is there still any problem of installation of graphviz on Mac and Windows? Could somebody with a winOS test this?

**Comment by ColCarroll** [comment]
1. *can we make this the repr for `Model`?* Probably!   I would wrap it in a `try/except` and fall back to the latex repr.  `graphviz` is easier to install now, but the binaries are still a bit annoying, and I would not want to support that as an installation requirement.
2. *Have you considered daft?* I thought I did, and remembered thinking it was a little ugly, but I looked again and it seems awesome!  The implementation here is not tied closely to `graphviz`, so I (or _someone_) could drop in another graphing engine reasonably easily.
3. *Is there an installation problem on Mac or Windows?*  I've run this on OSX and on Ubuntu.  I actually installed it on OSX using `brew install graphviz` and then `pip install graphviz`, but then verified that it worked in a conda environment, too.  No idea about windows :grimacing:

**Comment by twiecki** [comment]
This is awesome, the radon example is much improved due to this.

**Comment by fonnesbeck** [comment]
@ColCarroll fair enough on not wanting to have graphviz as a hard dependency. Would be cool, but probably not worth the pain.

**Comment by ColCarroll** [comment]
I took a small look at making this the repr, but it looks confusing (https://github.com/ipython/ipython/issues/11202). 

`daft` looks great, but requires manually specifying the position of each node.  If someone picked up https://github.com/dfm/daft/issues/81 then I think it would definitely be the right choice.  

Anyways, I think this is ready to merge.

**Comment by ColCarroll** [comment]
(merged to avoid conflicts with README.md with a bunch of other PRs coming... feel free to revert if there are more issues!)

**Pull request merged**